### ISSUE TITLE
putb(T msg, final long timeoutMillis)

### ISSUE THREAD

**Issue by CurryGaifan** [kilim/kilim#48]
Hi,
I read the MailBox.java and have a question. Why the method `putb(T msg, final long timeoutMillis)`  not reput the msg when the mailbox is not full. But the method `put(T msg, int timeoutMillis)`  reput the msg in a loop.
Thanks.

```
    /**
     * put a non-null message in the mailbox, and block the calling thread  for timeoutMillis
     * if the mailbox is full. 
     */
    public void putb(T msg, final long timeoutMillis) {
        BlockingSubscriber evs = new BlockingSubscriber();
        if (!put(msg, evs)) {
            evs.blockingWait(timeoutMillis);
        }
        if (!evs.eventRcvd) {
            removeSpaceAvailableListener(evs);
        }
    }
```

```
/**
     * put a non-null message in the mailbox, and pause the calling task  for timeoutMillis
     * if the mailbox is full. 
     */

    public boolean put(T msg, int timeoutMillis) throws Pausable {
        final Task t = Task.getCurrentTask();
        long begin = System.currentTimeMillis();
        while (!put(msg, t)) {
            TimerTask tt = new TimerTask() {
                public void run() {
                    Mailbox.this.removeSpaceAvailableListener(t);
                    t.onEvent(Mailbox.this, timedOut);
                }
            };
            Task.timer.schedule(tt, timeoutMillis);
            Task.pause(this);
            removeSpaceAvailableListener(t);
            if (System.currentTimeMillis() - begin >= timeoutMillis) {
                return false;
            }
        }
        return true;
    }
```

**Comment by kilim** [comment]
You may be right. I should look carefully to see why a loop is not needed. The intention is the same in both: attempt to do a put until we run out of time.

**Comment by nqzero** [comment]
note: the version of this code that i'm familiar with is from the perf
branch, and i haven't looked in depth at putb()

i believe the reason for the loop in the Pausable version is that the
Task.pause() is allowed to have a spurious resume. putb() uses wait/notify
(with a loop) and onEvent (which i don't believe is allowed to be spurious)

**Comment by zy416548283** [comment]
@kilim  @nqzero 
Here, we has similar question lastly.
```
The intention is the same in both: attempt to do a put until we run out of time.
```
But now, putb will not reput the msg into the mailbox;
In some situations, this behavior may be dangerous.

**Comment by nqzero** [comment]
@zy416548283 i'm just looking at this code in detail now. do you have an example that triggers the spurious return ? i'm assuming the right thing to do is wrap the put in a loop, but i'd like to understand the mechanism better first, so an example would help

if not, i'll just keep digging deeper into the code

**Comment by nqzero** [comment]
@zy416548283 i came up with a very simple test case, so don't worry about my last request for one

**Comment by nqzero** [comment]
this is fixed (for both putb and getb) in [kilim 2.0](https://github.com/nqzero/kilim)

@CurryGaifan thanks for noticing the inconsistency, and @zy416548283 thanks for explaining the significance to me
